src/basegraph.h

```diff
@@ -0,0 +1,355 @@
+				   std::vector<std::string>& inputs,
+				   std::vector<std::string>& outputs);
+
+	/**
+	 * \brief add layer / operator to the computational graph, reduced version
+	 * @param opname name of the layer
+	 * @param optype type of the layer
+	 * @return  added vertex
+	 */
+	Vertex add_layer(std::string opname, std::string optype);
+
+	/**
+	 * \brief set general input
+	 * @param name name of input data/blob
+	 * @param inputdim dimensions of the blob
+	 * @return the added vertex
+	 */
+	Vertex set_input(std::string name, std::vector<int>& inputdim);
+
+	/**
+	 * \brief add/bind some inputs to a vertex
+	 * if they already exist, they are bound, if not they are allocated
+	 * @param v vertex to add inputs to
+	 * @param inputs name of inputs
+	 * @return list of data inputs
+	 */
+	std::vector<Vertex> add_inputs(Vertex v, std::vector<std::string>& inputs);
+
+	/**
+	 * \brief add/bind some inputs to a vertex
+	 * if they already exist, they are bound, if not they are allocated
+	 * @param opname name of operator to add inputs to
+	 * @param inputs name of inputs
+	 * @return list of data inputs
+	 */
+	std::vector<Vertex> add_inputs(std::string opname, std::vector<std::string>& inputs);
+
+	/**
+	 * \brief add single input to a vertex, allocate data vertex if needed
+	 * @param v vertex to add input to
+	 * @param input name of data input
+	 * @return input vertex
+	 */
+	Vertex add_input(Vertex v, std::string input);
+
+
+	/**
+	 * \brief add single input to a vertex, allocate data vertex if needed
+	 * @param opname operator to add input to
+	 * @param input name of data input
+	 * @return input vertex
+	 */
+	Vertex add_input(std::string opname, std::string inputs);
+
+	/**
+	 * \brief add outputs to an operator vertex, allocates outputs if needed
+	 * @param v vertex to add ouput to
+	 * @param outputs name of data outputs
+	 * @return vector of data vertices
+	 */
+	std::vector<Vertex> add_outputs(Vertex v, std::vector<std::string>& outputs);
+
+
+	/**
+	 * \brief add outputs to an operator , allocates outputs if needed
+	 * @param opname name of operator to add ouput to
+	 * @param outputs name of data outputs
+	 * @return vector of data vertices
+	 */
+	std::vector<Vertex> add_outputs(std::string opname, std::vector<std::string>& outputs);
+
+
+	/**
+	 * \brief add output to an operator vertex, allocates output if needed
+	 * @param v vertex to add ouput to
+	 * @param varname name of data output
+	 * @return data vertex
+	 */
+	Vertex add_output(Vertex v, std::string varname);
+
+	/**
+	 * \brief add output to an operator , allocates output if needed
+	 * @param opnameto add ouput to
+	 * @param varname name of data output
+	 * @return data vertex
+	 */
+	Vertex add_output(std::string opname, std::string varname);
+
+	/**
+	 * \brief set loss to be used at learning
+	 * @param name  loss name
+	 */
+	void set_loss(std::string name) {_loss= name;}
+
+	/**
+	 * topological_sort of vertices, used internally
+	 */
+	void sort_all();
```

Comment:
"used internally" so it should be private?

Reply:
for now sort_all is used only by finalize(), but in my opinion there will be uses cases where finalize() will not do the job, and someone will want to sort explicitly then do other things things than what finalize() does